File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/filter/telegram/VenueFilterParser.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.telegram;

import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.NotNullFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.StringFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.LocationFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.telegram.VenueFilter;
import io.github.oleksivio.telegram.bot.api.model.objects.std.Venue;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.impl.UnionExtractValidator;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Optional;

@Component
public class VenueFilterParser implements FilterParser<VenueFilter, Venue> {

    @Override
    public Validator<Venue> createChecker(VenueFilter annotation, Finder finder) {
        UnionExtractValidator<Venue> unionExtractValidator = new UnionExtractValidator<>();

        Arrays.stream(annotation.validator())
                .map(validatorName -> finder.find(validatorName, Venue.class))
                .forEach(validator -> unionExtractValidator.add(Optional::of, validator));

        LocationFilter location = annotation.location();
        if (location.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getLocation()), finder.find(location));
        }
        StringFilter title = annotation.title();
        if (title.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getTitle()), finder.find(title));
        }
        StringFilter address = annotation.address();
        if (address.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getAddress()), finder.find(address));
        }
        NotNullFilter foursquareId = annotation.foursquareId();
        if (foursquareId.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getFoursquareId()), finder.find(foursquareId));
        }
        StringFilter foursquareType = annotation.foursquareType();
        if (foursquareType.status().isActive()) {
            unionExtractValidator.add(in -> Optional.ofNullable(in.getFoursquareType()), finder.find(foursquareType));
        }

        return unionExtractValidator;
    }

    @Override
    public Class<VenueFilter> getAnnotationClass() {
        return VenueFilter.class;
    }

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/annotations/parser/filter/telegram/VenueFilterParser.kt
package io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.filter.telegram

import io.github.oleksivio.tl.kbot.core.annotations.filter.composite.VenueFilter
import io.github.oleksivio.tl.kbot.server.api.objects.std.Venue
import io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.impl.UnionExtractValidator
 

  
class VenueFilterParser :
    FilterParser<VenueFilter, Venue> {

    override val parserAnnotationClass = VenueFilter::class

    override fun createChecker(annotation: VenueFilter, finder: Finder): Validator<Venue> {
        val unionExtractValidator =
            UnionExtractValidator<Venue>()

        annotation.validator
                .map { validatorName -> finder.find(validatorName, Venue::class) }
                .forEach { validator -> unionExtractValidator.add({ it }, validator) }

        val location = annotation.location
        if (location.status.isActive) {
            unionExtractValidator.add({ it.location }, finder.find(location))
        }
        val title = annotation.title
        if (title.status.isActive) {
            unionExtractValidator.add({ it.title }, finder.find(title))
        }
        val address = annotation.address
        if (address.status.isActive) {
            unionExtractValidator.add({ it.address }, finder.find(address))
        }
        val foursquareId = annotation.foursquareId
        if (foursquareId.status.isActive) {
            unionExtractValidator.add({ it.foursquareId }, finder.find(foursquareId))
        }
        val foursquareType = annotation.foursquareType
        if (foursquareType.status.isActive) {
            unionExtractValidator.add({ it.foursquareType }, finder.find(foursquareType))
        }

        return unionExtractValidator
    }

}

